src/engine/subscription/Subscription.ts

```diff
@@ -0,0 +1,311 @@
+        `Duplicate subscription name '${subscription.name}' found in '${entity.name}'`,
+    );
+
+    subscriptionNames.push(subscription.name);
+
+    if (subscription.attributes) {
+      passOrThrow(
+        (isArray(subscription.attributes, true) &&
+          subscription.type === SUBSCRIPTION_TYPE_CREATE) ||
+          isArray(subscription.attributes, false),
+        () =>
+          `Subscription '${entity.name}.${subscription.name}' needs to have a list of attributes`,
+      );
+
+      subscription.attributes.map(attribute => {
+        passOrThrow(
+          typeof attribute === 'string',
+          () =>
+            `Subscription '${entity.name}.${subscription.name}' needs to have a list of attribute names`,
+        );
+      });
+
+      passOrThrow(
+        subscription.attributes.length === uniq(subscription.attributes).length,
+        () =>
+          `Subscription '${entity.name}.${subscription.name}' needs to have a list of unique attribute names`,
+      );
+
+      subscription.attributes.map(attributeName => {
+        passOrThrow(
+          entityAttributes[attributeName],
+          () =>
+            `Cannot use attribute '${entity.name}.${attributeName}' in subscription '${entity.name}.${subscription.name}' as it does not exist`,
+        );
+      });
+
+      if (subscription.type === SUBSCRIPTION_TYPE_CREATE) {
+        const missingAttributeNames = requiredAttributeNames.filter(
+          requiredAttributeName => {
+            return !subscription.attributes.includes(requiredAttributeName);
+          },
+        );
+
+        passOrThrow(
+          missingAttributeNames.length === 0,
+          () =>
+            `Missing required attributes in subscription '${entity.name}.${
+              subscription.name
+            }' need to have a defaultValue() function: [ ${missingAttributeNames.join(
+              ', ',
+            )} ]`,
+        );
+      }
+    } else if (
+      subscription.type === SUBSCRIPTION_TYPE_CREATE ||
+      subscription.type === SUBSCRIPTION_TYPE_UPDATE
+    ) {
+      const nonSystemAttributeNames = [];
+
+      mapOverProperties(entityAttributes, (attribute, attributeName) => {
+        if (!attribute.isSystemAttribute) {
+          nonSystemAttributeNames.push(attributeName);
+        }
+      });
+
+      subscription.attributes = nonSystemAttributeNames;
+    }
+
+    if (subscription.delimiter) {
+      passOrThrow(
+        typeof subscription.delimiter === 'string',
+        () =>
+          `Subscription '${entity.name}.${subscription.name}' delimiter should be a string`,
+      );
+    } else {
+      subscription.delimiter = '/';
+    }
+
+    if (subscription.wildCard) {
+      // todo make a list of valid wildCards (#, +, *)
+      passOrThrow(
+        typeof subscription.wildCard === 'string',
+        () =>
+          `Subscription '${entity.name}.${subscription.name}' wildCard should be a string`,
+      );
+    }
+
+    if (subscription.pattern) {
```

Comment:
Here i still hesitate to use a pattern prop to define the topic chain based on entity attributes.
Anyway for now we can use preProcessor to compose our own route.

Reply:
I guess you can draw here on your experiences with pubsub